### ISSUE TITLE
Issues starting the application

### ISSUE THREAD

**Issue by JimboJones3101** [ChriD/node-raumserver#70]
Hey there,

I wanted to install the current NodeJS Version of Raumserver which seems to have worked. Nonetheless I'm facing issues starting the application.

What I have done so far:
1. Installed NodeJS via ' sudo apt-get install nodejs npm ' > Done
2. run ' node -v ' > v8.11.1
3. Installed the package in my homefolder via ' cd /home/openhabian/ ' and then ' ' npm install node-raumserver' > 

```
(node:29088) [DEP0022] DeprecationWarning: os.tmpDir() is deprecated. Use os.tmpdir() instead.
node-raumserver@0.1.6 node_modules/node-raumserver
├── config@1.31.0 (json5@1.0.1)
└── node-raumkernel@1.2.22 (query-string@4.3.4, xml2js@0.4.19, winston@2.4.4, winston-daily-rotate-file@1.7.2, bonjour@3.5.0, upnp-device-client@1.0.2, request@2.88.0, node-ssdp@3.3.0)
```

(Seems to me that installation got through although the above "warning"?!)

4. changed the port in ' default.json ' like this 

` "port" : 8088 `

as my openhab-instance is running on port 8080 ; furthermore tried to change the logging to verbose by 

`"loglevel" : 3 // 0=error, 1=warnings, 2=info, 3=verbose, 4=debug, 5=silly`

5. run ' npm start ' >

```
[10:48:30] openhabian@openHABianPi:~/node_modules/node-raumserver$ npm start
(node:31215) [DEP0022] DeprecationWarning: os.tmpDir() is deprecated. Use os.tmpdir() instead.

> node-raumserver@0.1.6 start /home/openhabian/node_modules/node-raumserver
> node raumserver.js

module.js:549
    throw err;
    ^

Error: Cannot find module 'query-string'
    at Function.Module._resolveFilename (module.js:547:15)
    at Function.Module._load (module.js:474:25)
    at Module.require (module.js:596:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/home/openhabian/node_modules/node-raumserver/lib/lib.raumserver.js:3:21)
    at Module._compile (module.js:652:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)

npm ERR! node-raumserver@0.1.6 start: `node raumserver.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the node-raumserver@0.1.6 start script.
npm ERR! This is most likely a problem with the node-raumserver package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node raumserver.js
npm ERR! You can get their info via:
npm ERR!     npm owner ls node-raumserver
npm ERR! There is likely additional logging output above.
npm ERR! System Linux 4.14.79-v7+
npm ERR! command "/usr/bin/node" "/usr/bin/npm" "start"
npm ERR! cwd /home/openhabian/node_modules/node-raumserver
npm ERR! node -v v8.11.1
npm ERR! npm -v 1.4.21
npm ERR! code ELIFECYCLE
npm WARN This failure might be due to the use of legacy binary "node"
npm WARN For further explanations, please read
/usr/share/doc/nodejs/README.Debian
 
npm ERR! Error: EACCES: permission denied, open 'npm-debug.log'
npm ERR!  { Error: EACCES: permission denied, open 'npm-debug.log'
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'open',
npm ERR!   path: 'npm-debug.log' }
npm ERR! 
npm ERR! Please try running this command again as root/Administrator.

npm ERR! System Linux 4.14.79-v7+
npm ERR! command "/usr/bin/node" "/usr/bin/npm" "start"
npm ERR! cwd /home/openhabian/node_modules/node-raumserver
npm ERR! node -v v8.11.1
npm ERR! npm -v 1.4.21
npm ERR! path npm-debug.log
npm ERR! syscall open
npm ERR! code EACCES
npm ERR! errno -13
npm ERR! stack Error: EACCES: permission denied, open 'npm-debug.log'
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /home/openhabian/node_modules/node-raumserver/npm-debug.log
npm ERR! not ok code 0
```

Now I'm somehow lost.  Just tried to do ' sudo npm start ' and got this far:

```
[10:56:05] openhabian@openHABianPi:~/node_modules/node-raumserver$ sudo npm start
(node:31279) [DEP0022] DeprecationWarning: os.tmpDir() is deprecated. Use os.tmpdir() instead.

> node-raumserver@0.1.6 start /home/openhabian/node_modules/node-raumserver
> node raumserver.js

module.js:549
    throw err;
    ^

Error: Cannot find module 'query-string'
    at Function.Module._resolveFilename (module.js:547:15)
    at Function.Module._load (module.js:474:25)
    at Module.require (module.js:596:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/home/openhabian/node_modules/node-raumserver/lib/lib.raumserver.js:3:21)
    at Module._compile (module.js:652:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)

npm ERR! node-raumserver@0.1.6 start: `node raumserver.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the node-raumserver@0.1.6 start script.
npm ERR! This is most likely a problem with the node-raumserver package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node raumserver.js
npm ERR! You can get their info via:
npm ERR!     npm owner ls node-raumserver
npm ERR! There is likely additional logging output above.
npm ERR! System Linux 4.14.79-v7+
npm ERR! command "/usr/bin/node" "/usr/bin/npm" "start"
npm ERR! cwd /home/openhabian/node_modules/node-raumserver
npm ERR! node -v v8.11.1
npm ERR! npm -v 1.4.21
npm ERR! code ELIFECYCLE
npm WARN This failure might be due to the use of legacy binary "node"
npm WARN For further explanations, please read
/usr/share/doc/nodejs/README.Debian
 
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /home/openhabian/node_modules/node-raumserver/npm-debug.log
npm ERR! not ok code 0
```

Any ideas of how to proceed? I don't get any kind of logging within file ' npm-debug.log ' which looks like this

```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/node', '/usr/bin/npm', 'start' ]
2 info using npm@1.4.21
3 info using node@v8.11.1
4 verbose run-script [ 'prestart', 'start', 'poststart' ]
5 info prestart node-raumserver@0.1.6
6 info start node-raumserver@0.1.6
7 verbose unsafe-perm in lifecycle true
8 info node-raumserver@0.1.6 Failed to exec start script
9 error node-raumserver@0.1.6 start: `node raumserver.js`
9 error Exit status 1
10 error Failed at the node-raumserver@0.1.6 start script.
10 error This is most likely a problem with the node-raumserver package,
10 error not with npm itself.
10 error Tell the author that this fails on your system:
10 error     node raumserver.js
10 error You can get their info via:
10 error     npm owner ls node-raumserver
10 error There is likely additional logging output above.
11 error System Linux 4.14.79-v7+
12 error command "/usr/bin/node" "/usr/bin/npm" "start"
13 error cwd /home/openhabian/node_modules/node-raumserver
14 error node -v v8.11.1
15 error npm -v 1.4.21
16 error code ELIFECYCLE
17 verbose exit [ 1, true ]
```

Thanks in advance

**Comment by ChriD** [comment]
Hmm.. seems that "query-string" module is missing.
Can you try to install it? 
`npm install -g query-string`

**Comment by JimboJones3101** [comment]
Hey ChriD,

thanks for your quick response.

By running your code I get the following message:

```
[11:45:00] openhabian@openHABianPi:~/node_modules/node-raumserver$ npm install -g query-string
(node:687) [DEP0022] DeprecationWarning: os.tmpDir() is deprecated. Use os.tmpdir() instead.
npm ERR! tar.unpack untar error /home/openhabian/.npm/query-string/6.2.0/package.tgz
npm ERR! Error: EACCES: permission denied, mkdir '/usr/local/lib/node_modules'
npm ERR!  { Error: EACCES: permission denied, mkdir '/usr/local/lib/node_modules'
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'mkdir',
npm ERR!   path: '/usr/local/lib/node_modules',
npm ERR!   fstream_type: 'Directory',
npm ERR!   fstream_path: '/usr/local/lib/node_modules/query-string',
npm ERR!   fstream_class: 'DirWriter',
npm ERR!   fstream_stack: 
npm ERR!    [ '/usr/lib/nodejs/fstream/lib/writer.js:171:25',
npm ERR!      '/usr/lib/nodejs/mkdirp/index.js:46:53',
npm ERR!      'FSReqWrap.oncomplete (fs.js:152:21)' ] }
npm ERR! 
npm ERR! Please try running this command again as root/Administrator.

npm ERR! System Linux 4.14.79-v7+
npm ERR! command "/usr/bin/node" "/usr/bin/npm" "install" "-g" "query-string"
npm ERR! cwd /home/openhabian/node_modules/node-raumserver
npm ERR! node -v v8.11.1
npm ERR! npm -v 1.4.21
npm ERR! path /usr/local/lib/node_modules
npm ERR! syscall mkdir
npm ERR! fstream_path /usr/local/lib/node_modules/query-string
npm ERR! fstream_type Directory
npm ERR! fstream_class DirWriter
npm ERR! code EACCES
npm ERR! errno -13
npm ERR! stack Error: EACCES: permission denied, mkdir '/usr/local/lib/node_modules'
npm ERR! fstream_stack /usr/lib/nodejs/fstream/lib/writer.js:171:25
npm ERR! fstream_stack /usr/lib/nodejs/mkdirp/index.js:46:53
npm ERR! fstream_stack FSReqWrap.oncomplete (fs.js:152:21)
npm ERR! Error: EACCES: permission denied, open 'npm-debug.log'
npm ERR!  { Error: EACCES: permission denied, open 'npm-debug.log'
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'open',
npm ERR!   path: 'npm-debug.log' }
npm ERR! 
npm ERR! Please try running this command again as root/Administrator.

npm ERR! System Linux 4.14.79-v7+
npm ERR! command "/usr/bin/node" "/usr/bin/npm" "install" "-g" "query-string"
npm ERR! cwd /home/openhabian/node_modules/node-raumserver
npm ERR! node -v v8.11.1
npm ERR! npm -v 1.4.21
npm ERR! path npm-debug.log
npm ERR! syscall open
npm ERR! code EACCES
npm ERR! errno -13
npm ERR! stack Error: EACCES: permission denied, open 'npm-debug.log'
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /home/openhabian/node_modules/node-raumserver/npm-debug.log
npm ERR! not ok code 0
```

I just tried `sudo npm install -g query-string` and it seemed to have worked:

```
[11:45:09] openhabian@openHABianPi:~/node_modules/node-raumserver$ sudo npm install -g query-string
[sudo] password for openhabian: 
(node:722) [DEP0022] DeprecationWarning: os.tmpDir() is deprecated. Use os.tmpdir() instead.
query-string@6.2.0 /usr/local/lib/node_modules/query-string
├── strict-uri-encode@2.0.0
└── decode-uri-component@0.2.0
```

Nonetheless when I try to run the application I get an error again:

```
[11:46:33] openhabian@openHABianPi:~/node_modules/node-raumserver$ npm start
(node:759) [DEP0022] DeprecationWarning: os.tmpDir() is deprecated. Use os.tmpdir() instead.

> node-raumserver@0.1.6 start /home/openhabian/node_modules/node-raumserver
> node raumserver.js

module.js:549
    throw err;
    ^

Error: Cannot find module 'query-string'
    at Function.Module._resolveFilename (module.js:547:15)
    at Function.Module._load (module.js:474:25)
    at Module.require (module.js:596:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/home/openhabian/node_modules/node-raumserver/lib/lib.raumserver.js:3:21)
    at Module._compile (module.js:652:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)

npm ERR! node-raumserver@0.1.6 start: `node raumserver.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the node-raumserver@0.1.6 start script.
npm ERR! This is most likely a problem with the node-raumserver package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node raumserver.js
npm ERR! You can get their info via:
npm ERR!     npm owner ls node-raumserver
npm ERR! There is likely additional logging output above.
npm ERR! System Linux 4.14.79-v7+
npm ERR! command "/usr/bin/node" "/usr/bin/npm" "start"
npm ERR! cwd /home/openhabian/node_modules/node-raumserver
npm ERR! node -v v8.11.1
npm ERR! npm -v 1.4.21
npm ERR! code ELIFECYCLE
npm WARN This failure might be due to the use of legacy binary "node"
npm WARN For further explanations, please read
/usr/share/doc/nodejs/README.Debian
 
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /home/openhabian/node_modules/node-raumserver/npm-debug.log
npm ERR! not ok code 0
```

I'm really sorry that it somethow is not working for me and thus I'm bothering you.... 
Can you assist further?

**Comment by ChriD** [comment]
Did you try staring as mentioned on the main side?
Move into the node-raumserver module directory and then do npm start

```
cd node_modules/node-raumserver
npm start
```

**Comment by JimboJones3101** [comment]
This is what I'm doing. I installed the whole package in

/home/openhabian/node_modules/node-raumserver

When I run the command I'm in that exact folder so I just have to run it without mentioning the whole part. 

Basically this is what the main page says. "create a folder with a name of your choice" > I have decided (of course I could change this if it would help) to store everything around this app in my users home directory. I would assume access rights should have no impact as the user running it "openhabian" has admin rights.

:-(

Any other idea?

**Comment by davie2000** [comment]
I have to start my raumserver.js with "sudo". Maybe it helps?

**Comment by ChriD** [comment]
I'm not sure what the problem is. It seems he cant find any module 
Path problems within node installation?

**Comment by JimboJones3101** [comment]
I don't get it :-(

```
[13:43:10] openhabian@openHABianPi:~/node_modules/node-raumserver$ sudo npm start
[sudo] password for openhabian: 
(node:2469) [DEP0022] DeprecationWarning: os.tmpDir() is deprecated. Use os.tmpdir() instead.

> node-raumserver@0.1.6 start /home/openhabian/node_modules/node-raumserver
> node raumserver.js

module.js:549
    throw err;
    ^

Error: Cannot find module 'query-string'
    at Function.Module._resolveFilename (module.js:547:15)
    at Function.Module._load (module.js:474:25)
    at Module.require (module.js:596:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/home/openhabian/node_modules/node-raumserver/lib/lib.raumserver.js:3:21)
    at Module._compile (module.js:652:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)

npm ERR! node-raumserver@0.1.6 start: `node raumserver.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the node-raumserver@0.1.6 start script.
npm ERR! This is most likely a problem with the node-raumserver package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node raumserver.js
npm ERR! You can get their info via:
npm ERR!     npm owner ls node-raumserver
npm ERR! There is likely additional logging output above.
npm ERR! System Linux 4.14.79-v7+
npm ERR! command "/usr/bin/node" "/usr/bin/npm" "start"
npm ERR! cwd /home/openhabian/node_modules/node-raumserver
npm ERR! node -v v8.11.1
npm ERR! npm -v 1.4.21
npm ERR! code ELIFECYCLE
npm WARN This failure might be due to the use of legacy binary "node"
npm WARN For further explanations, please read
/usr/share/doc/nodejs/README.Debian
 
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /home/openhabian/node_modules/node-raumserver/npm-debug.log
npm ERR! not ok code 0
```

> Path problems within node installation?

But what kind of problems might there be? I'm not sooo deep into the OS to think of anything....

**Comment by davie2000** [comment]
You are calling "sudo npm start" in your node-raumserver directory.

I do the following on Ubuntu 14:
```
davie2000@myVDR:~$ cd /home/davie2000/my-raumserver/node_modules/node-raumserver
davie2000@myVDR:~/my-raumserver/node_modules/node-raumserver$ sudo node raumserver.js
```

I think with "sudo npm start" only the start-command runs with sudo, but not the call "node raumserver.js". At least your log is telling me that: "npm ERR! node-raumserver@0.1.6 start: `node raumserver.js`".
I think that you have to use sudo because you installed query-string (or the whole npm) with sudo.
This is all "dangerous half knowledge" - but you could give it a try. Hope, it works!

**Comment by JimboJones3101** [comment]
Sooooo,

I seriously don't know what went wrong with my old installation, but I rolled back (I've got my RPi backed up every major version of my home automation system) the solution and installed it again and everything went fine. 

This time I used another approach to install nodeJS. Instead of (like described above) installing it via 

`sudo apt-get install nodejs npm`

I installed it via 

```
curl -sL https://deb.nodesource.com/setup_11.x | sudo -E bash -
sudo apt install -y nodejs
```

which resulted in version 11.4 instead of 8.11.1.

After this installation I did the same, thus creating a folder in my admins home folder and then following ChriD's guide via

```
npm install node-raumserver
cd ./node_modules/node-raumserver
npm start
```

This time it worked and I'm able to control my Raumfeld via HTTP requests. 
Thus this incident can be closed.

Thanks again for your support!
